ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -13,161 +13,304 @@
+   * @param channel The channel from which data needs to be written from
+   * @param size The amount of data in bytes to be written from the channel
+   * @throws IllegalArgumentException if the {@code size} is greater than a single segment's size.
+   * @throws IllegalStateException if there no more capacity in the log.
+   * @throws IOException if there was an I/O error while writing.
+   */
+  @Override
+  public void appendFrom(ReadableByteChannel channel, long size)
       throws IOException {
-    return fileChannel.size();
+    rollOverIfRequired(size);
+    activeSegment.appendFrom(channel, size);
   }
 
-  public void setLogEndOffset(long endOffset)
-      throws IOException {
-    long fileSize = fileChannel.size();
-    if (endOffset < 0 || endOffset > fileSize) {
-      throw new IllegalArgumentException("Log : " + file.getAbsolutePath() + " endOffset " + endOffset +
-          " outside the file size " + fileSize);
+  /**
+   * Sets the active segment in the log.
+   * @param name the name of the log segment that is to be marked active.
+   */
+  void setActiveSegment(String name) {
+    if (!segmentsByName.containsKey(name)) {
+      throw new IllegalArgumentException("There is no log segment with name: " + name);
     }
-    fileChannel.position(endOffset);
-    logger.trace("Log : {} setting log end offset {}", file.getAbsolutePath(), endOffset);
-    this.currentWriteOffset.set(endOffset);
+    activeSegment = segmentsByName.get(name);
   }
 
-  public long getLogEndOffset() {
-    return currentWriteOffset.get();
+  /**
+   * @return a segment iterator that can iterate over all the log segments in this log
+   */
+  Iterator<Map.Entry<String, LogSegment>> getSegmentIterator() {
+    return segmentsByName.entrySet().iterator();
   }
 
-  @Override
-  public int appendFrom(ByteBuffer buffer)
-      throws IOException {
-    if (currentWriteOffset.get() + buffer.remaining() > capacityInBytes) {
-      metrics.overflowWriteError.inc(1);
-      throw new IllegalArgumentException(
-          "Log : " + file.getAbsolutePath() + " error trying to append to log from buffer since new data size " +
-              buffer.remaining() + " exceeds total log size " + capacityInBytes);
+  /**
+   * @param name the name of the segment required.
+   * @return a {@link LogSegment} with {@code name} if it exists.
+   */
+  LogSegment getSegment(String name) {
+    return segmentsByName.get(name);
+  }
+
+  /**
+   * @return the end offset of the log abstraction.
+   */
+  Offset getLogEndOffset() {
+    LogSegment segment = activeSegment;
+    return new Offset(segment.getName(), segment.getEndOffset());
+  }
+
+  /**
+   * @return the currently capacity used of the log abstraction. Includes "wasted" space at the end of
+   * {@link LogSegment} instances that are not fully filled.
+   */
+  long getUsedCapacity() {
+    long usedCapacity = 0;
+    for (Map.Entry<String, LogSegment> entry : segmentsByName.entrySet()) {
```

Comment:
Yes - it is hard to keep this updated for compaction. Also the number of segments is finite and this loop will not take very long.